File: PAT_Advanced/1048.py
# PAT Advanced 1048 Find Coins
# 第4个测试点超时，不知道有没有AC的可能性
raw_N, raw_M = input().split()
raw_vector = input()
N = int(raw_N)
M = int(raw_M)
coins = list(map(int, raw_vector.strip().split()))
coins .sort()
prev = -1
flag_found = 0
for i in range(0, len(coins)):
    if prev == coins[i]:
        i = i + 1
    else:
        # 二分查找 i+1~len(coin)
        left = i + 1
        right = len(coins) - 1
        # 以“//”运算符得到int/int = int形式，否则商为float
        mid = (left + right) // 2
        while left <= right:
            if coins[mid] + coins[i] == M:
                flag_found = 1
                break
            if coins[mid] < M - coins[i]:
                left = mid + 1
            else:
                right = mid - 1
            mid = (left + right)//2
        if flag_found == 1:
            print("%d %d" % (coins[i], coins[mid]))
            break
        prev = coins[i]
if flag_found == 0:
    print("No Solution")

File: PAT/PAT Advanced/c++/1048.cpp
#include <iostream>
#include <string>
#include <cstdio>
#include <map>
using namespace std;

//PAT (Advanced Level) 1048 Find Coins

int M,N;
map<int,int> coins;

void input(){
    cin >> N >> M;
    for (int i = 0; i < N; ++i) {
        int coin;
        scanf("%d",&coin);
        if(coins.find(coin) == coins.end())
            coins[coin] = 1;
        else
            coins[coin]++;
    }
}

int checkCoinValue(int v1){
    int v2 = -1;
    if(v1 * 2 == M && coins[v1] > 1)
        return v1;
    if(coins.find(M - v1) != coins.end() && v1 * 2 != M )
        return M - v1;
    return v2;
}
int main(){
    input();
    for (auto it = coins.begin(); it != coins.end() ; ++it) {
        int v2 = checkCoinValue(it->first);
        if(v2 == -1)
            continue;
        else{
            cout << it->first << ' ' << v2;
            return 0;
        }
    }
    cout << "No Solution" << endl;
    return 0;
}
